src/extension.ts

```diff
@@ -164,3 +169,67 @@ async function ensureRuntimeDependencies(extension: vscode.Extension<object>, lo
         return true;
     }
 }
+
+class DocumentLinkProvider implements vscode.DocumentLinkProvider {
+    private _client: LanguageClient;
+
+    constructor(client: LanguageClient) {
+        this._client = client;
+    }
+
+    // Provides the text document with the ranges for document links
+    provideDocumentLinks(document: vscode.TextDocument, token: CancellationToken): vscode.ProviderResult<vscode.DocumentLink[]> {
+        return this._findDocLinks(document);
+    }
+
+    // Adds the target uri to the document link
+    async resolveDocumentLink(link: vscode.DocumentLink, token: CancellationToken): Promise<vscode.DocumentLink> {
+        const document = vscode.window.activeTextEditor.document;
+        const text = document.getText(link.range);
+        link.target = await this._findTargetUri(document, text);
+        return link;
+    }
+
+    // Returns document links found within the current text document
+    private _findDocLinks(document: vscode.TextDocument): vscode.DocumentLink[] {
+        const docText = document.getText();
+        const docRoles = docText.match(/:doc:`.+?`/gs);
+
+        if (docRoles === null) return [];
+
+        let doclinks: vscode.DocumentLink[] = [];
+        let docRoleOffsetStart = -1; // Initiated to -1 to accommodate 0th index
+
+        // For every doc role found, find their respective target
+        docRoles.forEach(async (docRole, index) => {
+            docRoleOffsetStart = docText.indexOf(docRole, docRoleOffsetStart + 1);
+
+            // Find target in doc role
+            const targetMatches = docRole.match(/((?<=<).+(?=>))|((?<=`)\S+(?!>)(?=`))/);
```

Comment:
This regex is a little scary. Unfortunately JS doesn't have a verbose regex syntax. Could you break this down into chunks and explain each part in comments?